src/Http/WebUtilities/src/FormPipeReader.cs

```diff
@@ -0,0 +1,342 @@
+                    !sequenceReader.IsNext(equalsDelimiter, true))
+                {
+                    break;
+                }
+
+                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> value, andDelimiter, false) ||
+                    !sequenceReader.IsNext(andDelimiter, true))
+                {
+                    if (!isFinalBlock)
+                    {
+                        break;
+                    }
+
+                    value = buffer.Slice(sequenceReader.Position);
+
+                    sequenceReader.Advance(value.Length);
+                }
+
+                // Need to call ToArray if the key/value spans multiple segments 
+                var decodedKey = GetDecodedStringFromReadOnlySequence(key);
+                var decodedValue = GetDecodedStringFromReadOnlySequence(value);
+
+                AppendAndVerify(ref accumulator, decodedKey, decodedValue);
+
+                consumed = sequenceReader.Position;
+            }
+
+            buffer = buffer.Slice(consumed);
+        }
+
+        private string GetDecodedStringFromReadOnlySequence(ReadOnlySequence<byte> ros)
+        {
+            if (ros.IsSingleSegment)
+            {
+                return GetDecodedString(ros.First.Span);
+            }
+
+            if (ros.Length < StackAllocThreshold)
+            {
+                Span<byte> buffer = stackalloc byte[(int)ros.Length];
+                ros.CopyTo(buffer);
+                return GetDecodedString(buffer);
+            }
+            else
+            {
+                var byteArray = ArrayPool<byte>.Shared.Rent((int)ros.Length);
+                Span<byte> buffer = byteArray.AsSpan(0, (int)ros.Length);
+
+                ros.CopyTo(buffer);
+                var decodedString = GetDecodedString(buffer);
+
+                ArrayPool<byte>.Shared.Return(byteArray);
+                return decodedString;
+            }
+        }
+
+        // Check that key/value constraints are met and appends value to accumulator.
+        private KeyValueAccumulator AppendAndVerify(ref KeyValueAccumulator accumulator, string decodedKey, string decodedValue)
+        {
+            if (decodedKey.Length > KeyLengthLimit)
+            {
+                throw new InvalidDataException($"Form key or value length limit {KeyLengthLimit} exceeded.");
+            }
+
+            if (decodedValue.Length > ValueLengthLimit)
+            {
+                throw new InvalidDataException($"Form key or value length limit {ValueLengthLimit} exceeded.");
+            }
+
+            accumulator.Append(decodedKey, decodedValue);
+
+            if (accumulator.ValueCount > ValueCountLimit)
+            {
+                throw new InvalidDataException($"Form value count limit {ValueCountLimit} exceeded.");
+            }
+
+            return accumulator;
+        }
+
+        private string GetDecodedString(in ReadOnlySpan<byte> readOnlySpan)
```

Comment:
Don't use in for ReadOnlySpan or span its only needed for ReadOnlySequence (since that's a big struct)